libnetdata/url/url.c

```diff
@@ -79,3 +79,128 @@ char *url_decode_r(char *to, char *url, size_t size) {
 
     return to;
 }
+
+/**
+ * Is request complete?
+ *
+ * The URL will be ready top be parsed when the last characters are the string "\r\n\r\n",
+ * this function checks the last characters of the message.
+ *
+ * @param begin is the first character of the sequence to analyse.
+ * @param end is the last character of the sequence
+ * @param length is the length of the total of bytes read, it is not the difference between end and begin.
+ *
+ * @return It returns HTTP_VALIDATION_OK when the message is complete and HTTP_VALIDATION_INCOMPLETE otherwise.
+ */
+inline HTTP_VALIDATION url_is_request_complete(char *begin,char *end,size_t length) {
+    //Message cannot be complete when first and last address are the same
+    if ( begin == end) {
+        return HTTP_VALIDATION_INCOMPLETE;
+    }
+
+    //We always set as begin the first character of the message, but we wanna check
+    //the last characters, so I am adjusting it case it is possible.
+    if ( length > 3  ) {
+        begin = end - 4;
+    }
+
+    //I wanna have a pair of '\r\n', so I will count them.
+    uint32_t counter = 0;
+    do {
+        if (*begin == '\r') {
+            begin++;
+            if ( begin == end )
+            {
+                break;
+            }
+
+            if (*begin == '\n')
+            {
+                counter++;
+            }
+        } else if (*begin == '\n') {
```

Comment:
Logic:
If I have the string "\n\nAB" I think the we will have `counter == 2`, is this possible?

Reply:
you are correct and this will be discarded later in the parser, because we do not have a valid request. The syntax of this function was based in the h2o/h2o web server parser request. 
Case we would follow the RFC like it was written, we would need to accept only '\r\n\r\n' at the end, but before the HTTP 1.1 some clients sent '\n\n' to finish a message, so it is a kind of compatibility. Now we have the script (request.sh) to test the URLs inside tests/urls, case you comment the line relative to (*begin == \n) you will see that wget can fail.

Reply:
If we are checking "\r\n\r\n" at the end, why can't we simply use `strncmp()`? Why do you need `begin`, `end`, and 'length', if only two of them are more than enough?